services/offline/src/lib/smart-interval.ts

```diff
@@ -0,0 +1,190 @@
+// todo: adjust defaults (e.g. 30 sec/5 min/1.5x)
+const FIVE_SECONDS = 5000
+const FIVE_MINUTES = 1000 * 60 * 5
+const DEFAULT_INCREMENT_FACTOR = 1.2
+const throwErrorIfNoCallbackIsProvided = (): void => {
+    throw new Error('Provide a callback')
+}
+
+// todo: remove console logs; though they are useful for testing
+
+class SmartInterval {
+    #initialDelay
+    #maxDelay
+    #delayIncrementFactor
+    #callback: () => void | boolean | Promise<void> | Promise<boolean>
+
+    #paused
+    #delay
+    // Timeout types are weird and initializing this to a dummy timeout
+    // clears up some checks
+    #timeout: NodeJS.Timeout = setTimeout(() => '', 0)
+    #standbyCallback: (() => void) | null = null
```

Comment:
Implementing this as a class with public and private fields works pretty well. You _could_ however also consider implementing this as a function i.e. `createSmartInterval`, which takes the same arguments, has pretty much the same internal implementation, and returns what are now the public methods. The main advantage of doing so is that you then don't have to worry about the scope of `this` within callbacks. I have only seen you use `.bind()` in one place, so perhaps it's not that relevant.

Reply:
I agree that this is a small thing since there's not much `this` management that needed to be done here, but I'll spend some time considering that refactor and what it would take, since no `this` management is better than a little.

I do think classes have some small benefits in being easy to read and easy to type

Reply:
After thinking it through, I realized this would be easy and elegant, so did the refactor in [this commit](https://github.com/dhis2/app-runtime/pull/1203/commits/e96e9c1b11aedc6b5962641c880f4b12898a8007) 🙂 Thanks for the suggestion!